src/main/java/org/tikv/common/PDClient.java

```diff
@@ -811,43 +819,60 @@ public String toString() {
     }
   }
 
-  private Metapb.Region decodeRegion(Metapb.Region region) {
+  @Override
+  public Metapb.Region decodeRegion(Metapb.Region region) {
     final boolean isRawRegion = conf.isRawKVMode();
     Metapb.Region.Builder builder =
         Metapb.Region.newBuilder()
             .setId(region.getId())
             .setRegionEpoch(region.getRegionEpoch())
             .addAllPeers(region.getPeersList());
 
-    if (region.getStartKey().isEmpty() || isRawRegion) {
+    if (region.getStartKey().isEmpty() || (conf.getApiVersion().isV1() && isRawRegion)) {
```

Comment:
We should have a fast path that simply return original region when it's unnecessary.